### ISSUE TITLE
Serialization Test Fails on Windows

### ISSUE THREAD

**Issue by theSundayProgrammer** [mlpack/mlpack#482]
This issue is similar to the load-save-test issue I raised earlier. But this time the error is in the test routine. Consider the following code from serial_test.cpp:
```
template<typename MatType,
         typename IArchiveType,
         typename OArchiveType>
void TestArmadilloSerialization(MatType& x)
{
  // First save it.
  ofstream ofs("test");
  OArchiveType o(ofs);
```

If the (I/0)ArchiveType is binary then the test fails due to the two character CR-LF eol marker. There are two options: one is to open the file as binary always; the other is partial function template specialization or something like that. I tried some tests using ios::binary and they seem to pass. But I am not sure about the general repercussions of opening a text file as binary.

**Comment by rcurtin** [comment]
This is definitely related to #468; I think it's actually the same issue, so I'll close this and comment there.

**Comment by theSundayProgrammer** [comment]
This is not a duplicate. In #468 there was nothing to do in mlpack. Here you know you are opening a binary file and yet you are not providing the ios::binary flag. Hence boost::serialization fails. My suggestion is to always provide the binary flag. If I do that in serial_test.cpp then all tests pass except DET Test.  Does that test pass on Linux/Unix?

**Comment by theSundayProgrammer** [comment]
I did a full rebuild and the test now passes

**Comment by rcurtin** [comment]
You're right, this isn't a duplicate.  Sorry about that.  I tested your commit and merged it in (c265e57).  I also made it so that when `data::Load()` loads models on Windows that are binary (`boost::archive::binary_iarchive` / `boost::archive::binary_oarchive`), binary mode is used (4ce2fc6).  All the tests pass on Linux, with the exception of DETTest, which is another bug I haven't managed to completely reproduce and fix (but that's outside the scope of this, and I'm hoping to devote a bit of time to that later today).

**Comment by theSundayProgrammer** [comment]
I ran  the test on Visual Studio with Boost Test Plugin. The DET test now
passes on my box. Even earlier if I create a separate test project with
only DET Test then it succeeds.
It could be because multiple tests are being run concurrently.


On Wed, Nov 25, 2015 at 5:55 AM, Ryan Curtin <notifications@github.com>
wrote:

> You're right, this isn't a duplicate. Sorry about that. I tested your
> commit and merged it in (c265e57
> <https://github.com/mlpack/mlpack/commit/c265e575099a7c1231c380c9de0f3320a892f1f4>).
> I also made it so that when data::Load() loads models on Windows that are
> binary (boost::archive::binary_iarchive / boost::archive::binary_oarchive),
> binary mode is used (4ce2fc6
> <https://github.com/mlpack/mlpack/commit/4ce2fc6557b8f05b871826d5aaf34013af6f2f8c>).
> All the tests pass on Linux, with the exception of DETTest, which is
> another bug I haven't managed to completely reproduce and fix (but that's
> outside the scope of this, and I'm hoping to devote a bit of time to that
> later today).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mlpack/mlpack/issues/482#issuecomment-159371702>.
>



-- 
Joseph Chakravarti Mariadassou
http://thesundayprogrammer.com
